### PR TITLE
Update TSP for markers

### PR DESCRIPTION
**Pull request by TheMatthew**

### CODE REVIEW

File: API.yaml [hunk 1]
```diff
@@ -774,6 +774,59 @@ paths:
             application/json:
               schema:
                 type: string
+  /experiments/{expUUID}/outputs/timeGraph/{outputID}/markers:
+    post:
+      tags:
+      - TimeGraph
+      summary: Tentative API for the TimeGraph views
+      operationId: getTimeGraphStates
```
Comment (bhufmann) [inline API.yaml]: Needs a different operationId

File: API.yaml [hunk 2]
```diff
@@ -774,6 +774,59 @@ paths:
...
+      description: Unique entry point for all TimeGraph states,
+        ensures that the same template is followed for all views
+      parameters:
+      - name: expUUID
```
Comment (bhufmann) [inline API.yaml]: Update description

File: API.yaml [hunk 3]
```diff
@@ -774,6 +774,59 @@ paths:
...
+        schema:
+          type: string
+      requestBody:
+        description: Query parameters to fetch the timegraph states.
+          Usually contains these parameters, but not limited to
+          - entries, array of all the entry (series) IDs
+          - nb, number of time stamps to sample
+          - startTime
+          - endTime
+          - labelRatio, ratio of state duration to label's string length to determine if the resolution is large enough to return the label
```
Comment (bhufmann) [inline API.yaml]: Is this valid for markers?

File: API.yaml [hunk 4]
```diff
@@ -774,6 +774,59 @@ paths:
...
+          description: Returns a list of time graph markers
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/GenericResponse'
+                  - type: object
+                    properties:
+                      model:
+                        $ref: '#/components/schemas/TimeGraphModel'
```
Comment (bhufmann) [inline API.yaml]: It will return a different model. TimeGraphMarker?

File: API.yaml [hunk 5]
```diff
@@ -1630,6 +1683,33 @@ components:
         styleKey:
           description: optional key to search for the style of this state. The style map can be obtained by using the style endpoint.
           type: string
+    TimeGraphMarker:
```
Comment (bhufmann) [inline API.yaml]: Do we need a category?

File: API.yaml [hunk 6]
```diff
@@ -1630,6 +1683,33 @@ components:
...
+          description: end time for this marker
+          type: integer
+          format: int64
+        label:
+          description: Text label to apply to this TimeGraphMarker if resolution permits.
+            Optional, no label is applied if absent.
+          type: string
+        tags:
+          description: Tags to apply on this marker
+          type: integer
```
Comment (bhufmann) [inline API.yaml]: Is this a bit mask?

File: API.yaml [hunk 7]
```diff
@@ -1630,6 +1683,33 @@ components:
...
+          format: int32
+        rows:
+          description: List of the time graph rows associated with this marker and zoom level.
```
Comment (bhufmann) [inline API.yaml]: do we need a list? Maybe

File: API.yaml [hunk 8]
```diff
@@ -1630,6 +1683,33 @@ components:
...
+          type: array
+          items:
+            $ref: '#/components/schemas/TimeGraphRow'
```
Comment (bhufmann) [inline API.yaml]: TimeGraphRow? Shouldn't it be the IDs?

File: API.yaml [hunk 9]
```diff
@@ -774,6 +774,59 @@ paths:
             application/json:
               schema:
                 type: string
+  /experiments/{expUUID}/outputs/timeGraph/{outputID}/markers:
```
Comment (delislesim) [inline API.yaml]: I think this should be renamed to _/experiments/{expUUID}/outputs/{outputID}/annotations:_

File: API.yaml [hunk 10]
```diff
@@ -489,6 +489,92 @@ paths:
             application/json:
               schema:
                 type: string
+ /experiments/{expUUID}/annotations:
```
Comment (delislesim) [inline API.yaml]: We should keep this endpoint, but it should be use to create annotations (bookmark). Right now it is used to fetch annotations, I think we can say that this endpoint is only to manage "global" annotations and each output have the responsibility to return these annotations when you query the specific annotations for this output.

File: API.yaml [hunk 11]
```diff
@@ -489,6 +489,92 @@ paths:
...
+    get:
```
Comment (delislesim) [inline API.yaml]: post

File: API.yaml [hunk 12]
```diff
@@ -489,6 +489,92 @@ paths:
...
+                type: array
+                items:
+                  $ref: '#/components/schemas/Bookmark'
+        404:
+          description: No such experiment
+          content:
+            application/json:
+              schema:
+                type: string
+/experiments/{expUUID}/annotations/{annotationId}:
```
Comment (delislesim) [inline API.yaml]: I don't really understand this endpoint, but I think what your are trying to do is the get annotation for a specific output. If this is the case I think we should use _/experiments/{expUUID}/outputs/XY/{outputID}/annotations:_ and _/experiments/{expUUID}/outputs/timeGraph/{outputID}/annotations:_ this way you don't have to specify a list of outputIds and having two endpoint gives the opportunity to specify a different model for XY and TimeGraph

File: API.yaml [hunk 13]
```diff
@@ -489,6 +489,92 @@ paths:
...
+    get:
```
Comment (delislesim) [inline API.yaml]: post

File: API.yaml [hunk 14]
```diff
@@ -489,6 +489,92 @@ paths:
...
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Query'
+      responses:
+        200:
+          description: Returns the queried annotations (an array of arrays of arrays , category -> outputID -> annotations)
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Annotations'
```
Comment (delislesim) [inline API.yaml]: This is not defined, Annotations should be generic and use _allOf_ to add specific fields for XY and TimeGraph

File: API.yaml [hunk 15]
```diff
@@ -489,6 +489,86 @@ paths:
...
+        description: ID of the output provider to query
+        required: true
+        schema:
+          type: string
+     requestBody:
+        description: Query parameters to fetch the annotations.
+          Usually contains these parameters, but not limited to
+          - list of outputId entries
+          - list of categoryID
+          - startTime
```
Comment (delislesim) [inline API.yaml]: This I'm not sure, I think we should send timeRequested as an array instead of start, end and resolution

File: API.yaml [hunk 16]
```diff
@@ -489,6 +489,86 @@ paths:
...
+      - name: outputID
+        in: path
+        description: ID of the output provider to query
+        required: true
+        schema:
+          type: string
+     requestBody:
+        description: Query parameters to fetch the annotations.
+          Usually contains these parameters, but not limited to
+          - list of outputId entries
```
Comment (delislesim) [inline API.yaml]: this is the same for all outputs, maybe this should be renamed to something else?

File: API.yaml [hunk 17]
```diff
@@ -489,6 +489,86 @@ paths:
...
+          - list of categoryID
+          - startTime
+          - endTime
+          - resolution
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Query'
+        required: true
+     responses:
```
Comment (delislesim) [inline API.yaml]: Is it suppose to be an Arrays of annotations? And the description need to be updated

File: API.yaml [hunk 18]
```diff
@@ -1430,6 +1510,50 @@ components:
           description: The bookmark's unique ID, generated by the server.
           type: integer
           format: int128
+    Annotation:
```
Comment (delislesim) [inline API.yaml]: Annotation need an entryId/seriesId and a y range in case of XY. Is it possible to generalize this concept?

File: API.yaml [hunk 19]
```diff
@@ -1430,6 +1510,50 @@ components:
...
+        label:
+          description: this annotation's name (Optional)
+          type: string
+        category:
+          description: The type of the annotation (generic, output, ...)
+          type: string
+        iconUrl:
+          description: URL to the annotation's icon (Optional)
+          type: string
+        UUID:
```
Comment (delislesim) [inline API.yaml]: Not major, but we just use _id_ usually

File: API.yaml [hunk 20]
```diff
@@ -1430,6 +1510,50 @@ components:
...
+          description: The annotation's unique ID, generated by the server.
+          type: integer
+          format: int128
+    AnnotationCategory:
+      type: object
+      properties:
+        label:
+          description: the name of the annotation category
+          type: string
+        tags:
```
Comment (delislesim) [inline API.yaml]: Not sure I understand this one.

File: API.yaml [hunk 21]
```diff
@@ -1430,6 +1510,50 @@ components:
...
+          description: The type of the annotation (generic, output, ...)
+          type: string
+        iconUrl:
+          description: URL to the annotation's icon (Optional)
+          type: string
+        UUID:
+          description: The annotation's unique ID, generated by the server.
+          type: integer
+          format: int128
+    AnnotationCategory:
```
Comment (delislesim) [inline API.yaml]: AnnotationCategory needs a recommended field or something like that to tell if this category should be shown by default

**Pull request closed without merging**